File: Leetcode/36 - Valid Sudoku/solution.py
def isValidSudoku(self, board: List[List[str]]) -> bool:
    for i in range(9):
        visited = [False] * 9
        for j in range(9):
            if board[i][j] == '.':
                continue
            num = ord(board[i][j]) - ord('1')
            if visited[num]:
                return False
            visited[num] = True

    for j in range(9):
        visited = [False] * 9
        for i in range(9):
            if board[i][j] == '.':
                continue
            num = ord(board[i][j]) - ord('1')
            if visited[num]:
                return False
            visited[num] = True

    for k in range(9):
        visited = [False] * 9
        currRow = (k // 3) * 3
        currCol = (k % 3) * 3
        for i in range(currRow, currRow + 3):
            for j in range(currCol, currCol + 3):
                if board[i][j] == '.':
                    continue
                num = ord(board[i][j]) - ord('1')
                if visited[num]:
                    return False
                visited[num] = True

    return True

File: Leetcode/36 - Valid Sudoku/solution.cpp
bool isValidSudoku(vector<vector<char>>& board) {
    for (int i = 0; i < 9; i++) {
        vector<bool> visited(9, false);
        for (int j = 0; j < 9; j++) {
            if (board[i][j] == '.') continue;
            int val = board[i][j] - '1';
            if (visited[val]) {
                return false;
            }
            visited[val] = true;
        }
    }

    for (int j = 0; j < 9; j++) {
        vector<bool> visited(9, false);
        for (int i = 0; i < 9; i++) {
            if (board[i][j] == '.') continue;
            int val = board[i][j] - '1';
            if (visited[val]) {
                return false;
            }
            visited[val] = true;
        }
    }

    for (int k = 0; k < 9; k++) {
        int currRow = (k / 3) * 3;
        int currCol = (k % 3) * 3;
        vector<bool> visited(9, false);
        for (int i = currRow; i < currRow + 3; i++) {
            for (int j = currCol; j < currCol + 3; j++) {
                if (board[i][j] == '.') continue;
                int val = board[i][j] - '1';
                if (visited[val]) {
                    return false;
                }
                visited[val] = true;
            }
        }
    }

    return true;
}